pkg/controllers/report/resource/controller.go

```diff
@@ -0,0 +1,247 @@
+}
+
+func (c *controller) GetResourceHash(uid types.UID) (Resource, schema.GroupVersionKind, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	for _, watcher := range c.dynamicWatchers {
+		if resource, exists := watcher.hashes[uid]; exists {
+			return resource, watcher.gvk, true
+		}
+	}
+	return Resource{}, schema.GroupVersionKind{}, false
+}
+
+func (c *controller) AddEventHandler(eventHandler EventHandler) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.eventHandlers = append(c.eventHandlers, eventHandler)
+	for _, watcher := range c.dynamicWatchers {
+		for uid, resource := range watcher.hashes {
+			eventHandler(uid, watcher.gvk, resource)
+		}
+	}
+}
+
+func (c *controller) updateDynamicWatchers() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	clusterPolicies, err := c.fetchClusterPolicies(logger)
+	if err != nil {
+		return err
+	}
+	policies, err := c.fetchPolicies(logger, metav1.NamespaceAll)
+	if err != nil {
+		return err
+	}
+	kinds := utils.BuildKindSet(logger, utils.RemoveNonBackgroundPolicies(logger, append(clusterPolicies, policies...)...)...)
+	gvrs := map[string]schema.GroupVersionResource{}
+	for _, kind := range kinds.List() {
+		gvr, err := c.client.Discovery().GetGVRFromKind(kind)
+		if err == nil {
+			gvrs[kind] = gvr
+		} else {
+			logger.Error(err, "failed to get gvr from kind", "kind", kind)
+		}
+	}
+	dynamicWatchers := map[schema.GroupVersionResource]*watcher{}
+	for kind, gvr := range gvrs {
+		// if we already have one, transfer it to the new map
+		if c.dynamicWatchers[gvr] != nil {
+			dynamicWatchers[gvr] = c.dynamicWatchers[gvr]
+			delete(c.dynamicWatchers, gvr)
+		} else {
+			logger.Info("start watcher ...", "gvr", gvr)
+			watchInterface, _ := c.client.GetDynamicInterface().Resource(gvr).Watch(context.TODO(), metav1.ListOptions{})
+			w := &watcher{
+				watcher: watchInterface,
+				gvk:     gvr.GroupVersion().WithKind(kind),
+				hashes:  map[types.UID]Resource{},
+			}
+			go func() {
+				gvr := gvr
+				defer logger.Info("watcher stopped")
+				for event := range watchInterface.ResultChan() {
+					switch event.Type {
+					case watch.Added:
+						c.updateHash(event.Object.(*unstructured.Unstructured), gvr)
+					case watch.Modified:
+						c.updateHash(event.Object.(*unstructured.Unstructured), gvr)
+					case watch.Deleted:
+						c.deleteHash(event.Object.(*unstructured.Unstructured), gvr)
+					}
+				}
+			}()
+			dynamicWatchers[gvr] = w
+			objs, _ := c.client.GetDynamicInterface().Resource(gvr).List(context.TODO(), metav1.ListOptions{})
+			for _, obj := range objs.Items {
+				uid := obj.GetUID()
+				hash := reportutils.CalculateResourceHash(obj)
+				w.hashes[uid] = Resource{
+					Hash:      hash,
+					Namespace: obj.GetNamespace(),
+					Name:      obj.GetName(),
+				}
+				c.notify(uid, w.gvk, w.hashes[uid])
+			}
+		}
+	}
+	// shutdown remaining watcher
+	for gvr, watcher := range c.dynamicWatchers {
+		watcher.watcher.Stop()
+		delete(c.dynamicWatchers, gvr)
```

Comment:
Do we need to delete this entry? L181 overwrites `c.dynamicWatchers` anyways.

Reply:
Doesn't hurt.